helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java

```diff
@@ -41,18 +60,279 @@
 public class WagedRebalancer implements GlobalRebalancer<ResourceControllerDataProvider> {
   private static final Logger LOG = LoggerFactory.getLogger(WagedRebalancer.class);
 
+  // When any of the following change happens, the rebalancer needs to do a global rebalance.
+  private static final Set<HelixConstants.ChangeType> GLOBAL_REBALANCE_REQUIRED_CHANGE_TYPES =
+      Collections.unmodifiableSet(new HashSet<>(Arrays
+          .asList(HelixConstants.ChangeType.RESOURCE_CONFIG, HelixConstants.ChangeType.CONFIG,
+              HelixConstants.ChangeType.INSTANCE_CONFIG)));
+
+  private ResourceChangeDetector _clusterDataDetector;
+  private AssignmentMetadataStore _assignmentMetadataStore;
+  private RebalanceAlgorithm _rebalanceAlgorithm;
+  private MappingCalculator<ResourceControllerDataProvider> _mappingCalculator;
+
   @Override
   public void init() {
+    // Use the mapping calculator in DelayedAutoRebalancer for calculating the final assignment output.
+    // TODO abstract and separate the mapping calculator logic from the DelayedAutoRebalancer
+    _mappingCalculator = new DelayedAutoRebalancer();
+    // TODO init the data detector accordingly, or change to final static method if possible.
+    _clusterDataDetector = new ResourceChangeDetector();
+    // TODO init the metadata store according to their requirement when integrate, or change to final static method if possible.
+    _assignmentMetadataStore = new AssignmentMetadataStore();
+    // TODO init the algorithm according to the requirement when integrate.
+    _rebalanceAlgorithm = new ConstraintsRebalanceAlgorithm();
   }
 
+  /**
+   * Calculate the new IdealStates of all the valid resources input.
+   *
+   * @param dataProvider       Cluster data cache that contains the current cluster status.
+   * @param resourceMap        A resource map that contains all the resources to be rebalanced.
+   *                           Note the resources that are not configured to use the WAGED rebalancer will be ignored.
+   * @param currentStateOutput The processed current states information.
+   * @return A map of rebalanced IdealState.
+   * @throws HelixException if the rebalance fails unexpectedly.
+   */
   @Override
-  public Map<String, IdealState> computeNewIdealStates(ResourceControllerDataProvider clusterData,
-      Map<String, Resource> resourceMap, final CurrentStateOutput currentStateOutput) {
-    return new HashMap<>();
+  public Map<String, IdealState> computeNewIdealStates(ResourceControllerDataProvider dataProvider,
```

Comment:
The public method and the related private methods are too long and unbelievably difficult to understand and maintain

Reply:
I have divided the PR to 2 pieces. But in general, this logic is one of the most complicated ones in our project. That's why we need to start early.
And, believe me, split them into sub methods won't help with readability...

Please try to understand what we are doing and comment with detail if you think any part can be simplified.